src/misc/coloring.c

```diff
@@ -129,6 +134,150 @@ static igraph_error_t igraph_i_vertex_coloring_greedy_cn(const igraph_t *graph,
+    for (igraph_integer_t nbr_indx = 0; nbr_indx < nbr_cnt; nbr_indx++) {
+        igraph_integer_t nbr = VECTOR(*neighbours)[nbr_indx];
+        if (!igraph_gen2wheap_has_elem(node_degrees_heap, nbr)) {
+            continue;
+        }
+        color_metadata = *((dsatur_t*)igraph_gen2wheap_get(node_degrees_heap, nbr));
+        if (!is_color_used_by_neighbour(colors, color, igraph_adjlist_get(adjlist, nbr))) {
+            color_metadata.saturation_degree++;
+        }
+        color_metadata.edge_degree--;
+        igraph_gen2wheap_modify(node_degrees_heap, nbr, &color_metadata);
+    }
+}
+
+static igraph_integer_t igraph_i_set_first_color(igraph_set_t *viable_colors) {
+    if (!igraph_set_contains(viable_colors, 0)) {
+        return 0;
+    }
+    igraph_integer_t state = 0, prev_element, next_element;
+    igraph_set_iterate(viable_colors, &state, &prev_element);
+    while (igraph_set_iterate(viable_colors, &state, &next_element)) {
+        if (next_element != prev_element + 1) {
+            return prev_element + 1;
+        }
+        prev_element = next_element;
+    }
+    return prev_element + 1;
+}
+
+static igraph_error_t igraph_i_dsatur_main_loop(
+    const igraph_t *graph, igraph_vector_int_t *colors,
+    igraph_gen2wheap_t *node_degrees_heap, igraph_adjlist_t *adjlist,
+    igraph_integer_t vc
+) {
+    igraph_integer_t vertices_colored = 0;
+    igraph_set_t viable_colors;
+    igraph_set_init(&viable_colors, 0);
+    IGRAPH_FINALLY(igraph_set_destroy, &viable_colors);
+    IGRAPH_CHECK(igraph_set_reserve(&viable_colors, vc));
+
+    while (vertices_colored < vc) {
+        igraph_integer_t node_to_color = igraph_gen2wheap_max_index(node_degrees_heap);
+        igraph_vector_int_t *neighbours = igraph_adjlist_get(adjlist, node_to_color);
+        igraph_integer_t neighbours_count = igraph_vector_int_size(neighbours);
+        igraph_set_clear(&viable_colors);
+        for (igraph_integer_t neighbor_index = 0 ; neighbor_index < neighbours_count ; neighbor_index++) {
+            igraph_integer_t neighbour = VECTOR(*neighbours)[neighbor_index];
+            if (VECTOR(*colors)[neighbour] != -1) {
+                igraph_set_add(&viable_colors, VECTOR(*colors)[neighbour]);
+            }
+        }
+        igraph_integer_t color = igraph_i_set_first_color(&viable_colors);
+        igraph_i_dsatur_update_heap(adjlist, node_degrees_heap, neighbours, colors, color);
+        VECTOR(*colors)[node_to_color] = color;
+
+        vertices_colored++;
+        IGRAPH_ALLOW_INTERRUPTION();
+    }
+    igraph_set_destroy(&viable_colors);
+    IGRAPH_FINALLY_CLEAN(1);
+    return IGRAPH_SUCCESS;
+}
+
+static igraph_error_t igraph_i_vertex_coloring_dsatur(
+    const igraph_t *graph, igraph_vector_int_t *colors
+) {
+    igraph_integer_t vc = igraph_vcount(graph);
+
+    if (vc == 0) {
+        return IGRAPH_SUCCESS ;
+    }
+
+    IGRAPH_CHECK(igraph_vector_int_resize(colors, vc));
+
```

Comment:
```suggestion
    IGRAPH_CHECK(igraph_vector_int_resize(colors, vc));

    if (vc == 0) {
        return IGRAPH_SUCCESS ;
    }
```

Otherwise the size of the `colors` vector won't be set correctly for the null graph.